### PR TITLE
feat: Add a separate function for service startup validation

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Also sorted module-private methods and aligned service url parsing logic

### CODE REVIEW

File: appium/webdriver/appium_service.py [hunk 1]
```diff
@@ -215,38 +119,44 @@ def start(self, **kwargs: Any) -> sp.Popen:
...
+            was_running = True
             self._process.terminate()
-            self._process.communicate(timeout=5)
-            is_terminated = True
+            try:
+                self._process.communicate(timeout=timeout)
+            except sp.SubprocessError:
+                import signal
+
+                os.kill(self._process.pid, signal.SIGKILL)
```
Comment (KazuCocoa) [inline appium/webdriver/appium_service.py:150]: Would `self._process.kill()` be better for cross platform?
https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

Comment (mykola-mokhnach) [reply appium/webdriver/appium_service.py:150]: process.kill is just an alias to process.terminate, which is already called above

Comment (KazuCocoa) [reply appium/webdriver/appium_service.py:150]: for Win? i though for Linux are macOS are sigkill in `kill` and alias to terminate for Win. So in terms of Win, yea, it's just an alias already called, so it calls  terminal twice thing

Comment (KazuCocoa) [reply appium/webdriver/appium_service.py:150]: Is it in `if _mswindows:`?

Screenshot 2024-10-03 at 10 43 08 AM

Comment (mykola-mokhnach) [reply appium/webdriver/appium_service.py]: I did not pay attention to the `else:` block. Changed to use .kill

File: appium/webdriver/appium_service.py [hunk 2]
```diff
@@ -215,38 +119,42 @@ def start(self, **kwargs: Any) -> sp.Popen:
...
         if self.is_running:
             assert self._process
+            was_running = True
             self._process.terminate()
-            self._process.communicate(timeout=5)
-            is_terminated = True
+            try:
+                self._process.communicate(timeout=timeout)
+            except sp.SubprocessError:
+                self._process.kill()
```
Comment (KazuCocoa) [inline appium/webdriver/appium_service.py:148]: actually, after kill, we need wait/communicate
```suggestion
                self._process.kill()
                _, _ = self._process. communicate(timeout=timeout)
```
(or print the error) to ensure the process state

**Review by KazuCocoa — APPROVED** [review]
left process handling comment, otherwise lg entirely

File: appium/webdriver/appium_service.py (hunk 2 above)
Comment (mykola-mokhnach) [reply appium/webdriver/appium_service.py:148]: kill sends SIGKILL, which terminates the process unconditionally.
maybe there is some official link to docs about the best way of cross-platform graceful/forceful process termination ?

Comment (KazuCocoa) [reply appium/webdriver/appium_service.py:148]: I haven't found out cross platform thing, but my comment to put communicate comes from https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate 's example code. Actually it is not after terminate's exception though.

On Windows, I explicitly called taskkill below with force option when terminate/kill fails.

```
os.system("taskkill /f /pid {}".format(process.pid))
process.wait(timeout=5)
```

Based on my past Win's experience in work, the taskkill helped to kill process forcefully. The terminate/kill in Python was a graceful one. Probably... this taskkill could be ok to add when we get a new issue though

Comment (mykola-mokhnach) [reply appium/webdriver/appium_service.py:148]: it's actually a good idea to use taskkill, I'll add it to the implementation

**Pull request merged**